src/Microsoft.Azure.SignalR.Common/ServiceConnections/RawMultiEndpointServiceConnectionContainer.cs

```diff
@@ -0,0 +1,167 @@
+        }
+
+        public async Task<bool> WriteAckableMessageAsync(ServiceMessage serviceMessage, CancellationToken cancellationToken = default)
+        {
+            // If we have multiple endpoints, we should wait to one of the following conditions hit
+            // 1. One endpoint responses "OK" state
+            // 2. All the endpoints response failed state including "NotFound", "Timeout" and waiting response to timeout
+            var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            var writeMessageTask = WriteMultiEndpointMessageAsync(serviceMessage, async connection =>
+            {
+                var succeeded = await connection.WriteAckableMessageAsync(serviceMessage, cancellationToken);
+                if (succeeded)
+                {
+                    tcs.TrySetResult(true);
+                }
+            });
+
+            // If tcs.Task completes, one Endpoint responses "OK" state.
+            var task = await Task.WhenAny(tcs.Task, writeMessageTask);
+
+            // This will throw exceptions in tasks if exceptions exist
+            await task;
+
+            return tcs.Task.IsCompleted;
+        }
+
+        private Task WriteMultiEndpointMessageAsync(ServiceMessage serviceMessage, Func<IServiceConnectionContainer, Task> inner)
+        {
+            var routed = TargetEndpoints?
+                .Select(endpoint =>
+                {
+                    var connection = endpoint?.ConnectionContainer;
+                    if (connection == null)
+                    {
+                        Log.EndpointNotExists(_logger, endpoint.ToString());
+                    }
+                    return (e: endpoint, c: connection);
+                })
+                .Where(c => c.c != null)
+                .Select(async s =>
+                {
+                    try
+                    {
+                        Log.RouteMessageToServiceEndpoint(_logger, serviceMessage, s.e.ToString());
+                        await inner(s.c);
+                    }
+                    catch (ServiceConnectionNotActiveException)
+                    {
+                        // log and don't stop other endpoints
+                        Log.FailedWritingMessageToEndpoint(_logger, serviceMessage.GetType().Name, (serviceMessage as IMessageWithTracingId)?.TracingId, s.e.ToString());
+                    }
+                }).ToArray();
+
+            if (routed == null || routed.Length == 0)
+            {
+                // check if the router returns any endpoint
+                Log.NoEndpointRouted(_logger, serviceMessage.GetType().Name);
+                return Task.CompletedTask;
+            }
+
+            if (routed.Length == 1)
+            {
+                return routed[0];
+            }
+
+            return Task.WhenAll(routed);
+        }
+
+        public Task StartAsync() => Task.CompletedTask;
```

Comment:
I feel implementing `StartAsync` and `StopAsync` makes it a real container, any reason not have it?

Reply:
Is it just some multiple endpoint message writer?

Reply:
Yes, it is just some multiple endpoint message writer